src/cascadia/TerminalSettingsEditor/Appearances.cpp

```diff
@@ -19,6 +20,20 @@ using namespace winrt::Windows::Foundation;
 using namespace winrt::Windows::Foundation::Collections;
 using namespace winrt::Microsoft::Terminal::Settings::Model;
 
+static std::unordered_set<std::wstring_view> DefaultFeatures{
+    L"rlig",
+    L"locl",
+    L"ccmp",
+    L"calt",
+    L"liga",
+    L"clig",
+    L"rnrn",
+    L"kern",
+    L"mark",
+    L"mkmk",
+    L"dist"
+};
```

Comment:
> std::array<std::wstring_view, 11>

I did think about this, but wouldn't this cause lookup to be slow? I figured the longer initialization would be better than having to loop through everytime for lookup

Reply:
A modern CPU can loop through like, a hundred string somparisons in an amount of time so infinitesimally small that it is _embarrassing_. For these 11 things, a flat array is perfectly fine. :)

Reply:
With `vcpkg` I always keep `google/benchmark` ready on my PC for these situations. 😄

<details>
<summary>Source Code</summary>

```cpp
#include <array>
#include <unordered_set>

#include <benchmark/benchmark.h>

#include <dwrite_3.h>

static constexpr std::array<std::wstring_view, 64> all_features{L"afrc", L"c2pc", L"c2sc", L"calt", L"case", L"ccmp", L"clig", L"cpsp", L"cswh", L"curs", L"dflt", L"dlig", L"expt", L"frac", L"fwid", L"half", L"haln", L"halt", L"hist", L"hkna", L"hlig", L"hwid", L"hojo", L"jp04", L"jp78", L"jp83", L"jp90", L"kern", L"liga", L"lnum", L"locl", L"mark", L"mgrk", L"mkmk", L"nalt", L"nlck", L"onum", L"ordn", L"palt", L"pcap", L"pnum", L"pwid", L"qwid", L"rlig", L"ruby", L"salt", L"sinf", L"smcp", L"smpl", L"ss01", L"ss02", L"ss03", L"subs", L"sups", L"swsh", L"titl", L"tnam", L"tnum", L"trad", L"twid", L"unic", L"vert", L"vrt2", L"zero"};

static std::unordered_set<std::wstring_view> features_set{L"calt", L"ccmp", L"clig", L"dist", L"kern", L"liga", L"locl", L"mark", L"mkmk", L"rlig", L"rnrn"};
static constexpr std::array<std::wstring_view, 11> features_array{L"calt", L"ccmp", L"clig", L"dist", L"kern", L"liga", L"locl", L"mark", L"mkmk", L"rlig", L"rnrn"};
static constexpr std::array features_array_plain{L"calt", L"ccmp", L"clig", L"dist", L"kern", L"liga", L"locl", L"mark", L"mkmk", L"rlig", L"rnrn"};
static constexpr std::array features_array_tags{DWRITE_MAKE_OPENTYPE_TAG('c', 'a', 'l', 't'), DWRITE_MAKE_OPENTYPE_TAG('c', 'c', 'm', 'p'), DWRITE_MAKE_OPENTYPE_TAG('c', 'l', 'i', 'g'), DWRITE_MAKE_OPENTYPE_TAG('d', 'i', 's', 't'), DWRITE_MAKE_OPENTYPE_TAG('k', 'e', 'r', 'n'), DWRITE_MAKE_OPENTYPE_TAG('l', 'i', 'g', 'a'), DWRITE_MAKE_OPENTYPE_TAG('l', 'o', 'c', 'l'), DWRITE_MAKE_OPENTYPE_TAG('m', 'a', 'r', 'k'), DWRITE_MAKE_OPENTYPE_TAG('m', 'k', 'm', 'k'), DWRITE_MAKE_OPENTYPE_TAG('r', 'l', 'i', 'g'), DWRITE_MAKE_OPENTYPE_TAG('r', 'n', 'r', 'n')};

static volatile size_t initial_idx = UINT64_C(1442695040888963407);

static void BM_unordered_set(benchmark::State& state) {
    size_t idx = initial_idx;
    for (auto _ : state) {
        const auto& feature = all_features[idx % all_features.size()];

        const bool exists = features_set.contains(feature);

        benchmark::DoNotOptimize(exists);
        idx = idx * UINT64_C(6364136223846793005) + UINT64_C(1442695040888963407);
    }
}
BENCHMARK(BM_unordered_set);

static void BM_array(benchmark::State& state) {
    size_t idx = initial_idx;
    for (auto _ : state) {
        const auto& feature = all_features[idx % all_features.size()];

        const auto exists = std::ranges::find(features_array, feature) != features_array.end();

        benchmark::DoNotOptimize(exists);
        idx = idx * UINT64_C(6364136223846793005) + UINT64_C(1442695040888963407);
    }
}
BENCHMARK(BM_array);

static void BM_array_plain(benchmark::State& state) {
    size_t idx = initial_idx;
    for (auto _ : state) {
        const auto& feature = all_features[idx % all_features.size()];

        bool exists = false;
        for (const auto f : features_array_plain) {
            if (wcscmp(feature.data(), f) == 0) {
                exists = true;
                break;
            }
        }

        benchmark::DoNotOptimize(exists);
        idx = idx * UINT64_C(6364136223846793005) + UINT64_C(1442695040888963407);
    }
}
BENCHMARK(BM_array_plain);

static void BM_array_tags(benchmark::State& state) {
    size_t idx = initial_idx;
    for (auto _ : state) {
        const auto& feature = all_features[idx % all_features.size()];

        const auto tag = DWRITE_MAKE_OPENTYPE_TAG(feature[0], feature[1], feature[2], feature[3]);
        const auto exists = std::ranges::find(features_array_tags, tag) != features_array_tags.end();

        benchmark::DoNotOptimize(exists);
        idx = idx * UINT64_C(6364136223846793005) + UINT64_C(1442695040888963407);
    }
}
BENCHMARK(BM_array_tags);

BENCHMARK_MAIN();
```

</details>

It's effectively a rounding error between each of these:
```
-----------------------------------------------------------
Benchmark                 Time             CPU   Iterations
-----------------------------------------------------------
BM_unordered_set       6.12 ns         6.14 ns    112000000   <-- std::unordered_set
BM_array               9.19 ns         9.21 ns     74666667   <-- std::array + string_view
BM_array_plain         9.03 ns         9.00 ns     74666667   <-- plain C, like in the 90s
BM_array_tags          3.62 ns         3.60 ns    186666667   <-- smort. transforms the string to a `DWRITE_FONT_FEATURE_TAG` first and then searches it
```

(The difference between 6ns and 9ns is about that of 1 function call to a function that has an empty body. To put that into perspective, creating a `winrt::hstring` is roughly 100x more expensive.)

They're all so similar even though they're O(1), O(n^2), O(n^2), and O(n) respectively simply because CPUs are really good at pipelining and caching and the O notation often only matters when your N is sufficiently large. In our case with N=11, even old CPUs will have an easy time to ~paper~ pipeline over any differences.

That's also why I don't consider this a pressing issue. I only said it because I think code is really pretty when it doesn't need many resources. The `unordered_set` is the only one out of the above that has any additional cost associated with it.
The way I see it, these tiny craftsmanship things have sort of a compounding effect. It's not a single component in Teams or whatever that makes it such a pain to work with, rather it's 100k cuts all over the place throughout their entire tech stack. It's very similar for Windows Terminal and its launch performance. When you look at the CPU usage in a timeline trace or in WPA, it looks like someone threw tiny confetti all over the place. 😄

Reply:
The craftsmanship angle for me is: I don't think that loading this library into our process should suddenly start costing more because it has to run another static initializer and make a bunch of heap allocations. 😄

Static constexpr _arrays of stuff_ are _dead_. Provably dead. They don't require us to run code.

I'll link back to the heroic savings we got when we replaced CodepointWidthDetector from an `unordered_set` to... well, a constexpr array of stuff: #2368 - Read the description for the fun!

Reply:
Leonard's measurement doesn't include constructing the unordered set, only _using_ it. The cost of constructing it is paid when you open the settings UI, and it definitely may be _small_, but it is not free.

Reply:
2 couple extras for reference:
```
-----------------------------------------------------------
Benchmark                 Time             CPU   Iterations
-----------------------------------------------------------
BM_unordered_set_flat  4.96 ns         5.00 ns    100000000   <-- til::linear_flat_set + til::hash
BM_unordered_set_fast  2.51 ns         2.51 ns    280000000   <-- til::linear_flat_set + the DWRITE_FONT_FEATURE_TAG trick
```